opennav_following/src/following_server.cpp

```diff
@@ -0,0 +1,749 @@
+      checkAndWarnIfPreempted<FollowObject>(following_action_server_, "follow_object"))
+    {
+      return false;
+    }
+
+    // Get the tracking pose from topic or frame
+    getTrackingPose(object_pose, target_frame);
+
+    // Get the pose at the distance we want to maintain from the object
+    // Stop and report success if goal is reached
+    auto target_pose = getPoseAtDistance(object_pose, desired_distance_);
+    if (isGoalReached(target_pose)) {
+      return true;
+    }
+
+    // The control law can get jittery when close to the end when atan2's can explode.
+    // Thus, we reduce the desired distance by a small amount so that the robot never
+    // gets to the end of the spiral before its at the desired distance to stop the
+    // following procedure.
+    const double backward_projection = 0.25;
+    const double effective_distance = desired_distance_ - backward_projection;
+    target_pose = getPoseAtDistance(object_pose, effective_distance);
+
+    // ... and transform the target_pose into base_frame
+    try {
+      tf2_buffer_->transform(
+        target_pose, target_pose, base_frame_, tf2::durationFromSec(transform_tolerance_));
+    } catch (const tf2::TransformException & ex) {
+      RCLCPP_WARN(get_logger(), "Failed to transform target pose: %s", ex.what());
+      return false;
+    }
+
+    // If the object is behind the robot, we reverse the control
+    geometry_msgs::msg::PoseStamped robot_pose;
+    if (!nav2_util::getCurrentPose(
+        robot_pose, *tf2_buffer_, target_pose.header.frame_id, base_frame_, transform_tolerance_,
+        iteration_start_time_))
+    {
+      RCLCPP_WARN(get_logger(), "Failed to get current robot pose");
+      return false;
+    }
+
+    // Compute and publish controls
+    auto command = std::make_unique<geometry_msgs::msg::TwistStamped>();
+    command->header.stamp = now();
+    if (!controller_->computeVelocityCommand(target_pose.pose, command->twist, true, false)) {
+      throw opennav_docking_core::FailedToControl("Failed to get control");
+    }
+    vel_publisher_->publish(std::move(command));
+
+    loop_rate.sleep();
+  }
+  return false;
+}
+
+bool FollowingServer::rotateToObject(
+  geometry_msgs::msg::PoseStamped & object_pose, const std::string & target_frame)
+{
+  const double dt = 1.0 / controller_frequency_;
+
+  // Compute initial robot heading
+  geometry_msgs::msg::PoseStamped robot_pose;
+  if (!nav2_util::getCurrentPose(
+      robot_pose, *tf2_buffer_, object_pose.header.frame_id, base_frame_, transform_tolerance_,
+      iteration_start_time_))
+  {
+    RCLCPP_WARN(get_logger(), "Failed to get current robot pose");
+    return false;
+  }
+  double initial_yaw = tf2::getYaw(robot_pose.pose.orientation);
+
+  // Search angles: 90 deg left, then sweep 180 deg to the right (ends 90 deg right of initial)
+  std::vector<double> angles = {initial_yaw + M_PI_2, initial_yaw - M_PI_2};
```

Comment:
Last change: maybe we should parameterize this sweep angle

Did you also test in simulation this works?